donkeycar/parts/lidar.py

```diff
@@ -13,6 +16,257 @@
+                        #       (faster than the scan rate of the lidar),
+                        #       then the returned scans will have some new
+                        #       values and some values that may have been
+                        #       seen in the previous scan.  The scan:index
+                        #       pair can be used to
+                        #       to 'diff' scans to see which measurements
+                        #       are are 'new' and which measurements are
+                        #       shared between scans.
+                        #
+                        #       The time at which the measurement was
+                        #       aquired is also included. In a moving
+                        #       vehicle older measurements are less
+                        #       relevant; the time field can be used to
+                        #       filter out older measurements or to
+                        #       visualize them differently (fading them out
+                        #       perhaps). It may also help when using a
+                        #       kinematic model to adjust for movement
+                        #       of the lidar when attached to a vehicle.
+                        #
+                        measurement = (distance, angle, now,
+                                        self.full_scan_count, self.full_scan_index)
+                        
+                        # grow buffer if necessary, otherwise overwrite
+                        if self.measurement_index >= len(self.measurements):
+                            self.measurements.append(measurement)
+                            self.measurement_count = self.measurement_index + 1
+                        else:
+                            self.measurements[self.measurement_index] = measurement  # noqa
+                        self.measurement_index += 1
+                        self.full_scan_index += 1
+                            
+            except serial.serialutil.SerialException:
+                logger.info('SerialException from RPLidar.')
+
+    def update(self):
+        start_time = time.time()
+        while self.running:
+            self.poll()
+            time.sleep(0)  # yield time to other threads
+        total_time = time.time() - start_time
+        scan_rate = self.full_scan_count / total_time
+        measurement_rate = self.total_measurements / total_time
+        logger.info("RPLidar total scan time = {time} seconds".format(time=total_time))
+        logger.info("RPLidar total scan count = {count} scans".format(count=self.full_scan_count))
+        logger.info("RPLidar total measurement count = {count} measurements".format(count=self.total_measurements))
+        logger.info("RPLidar rate = {rate} scans per second".format(rate=scan_rate))
+        logger.info("RPLidar rate = {rate} measurements per second".format(rate=measurement_rate))
```

Comment:
We generally switched to f-strings for readability and brevity (they also perform better, but I think this is less of a concern).

Reply:
You are right, but I'll leave it for now.